lib/widgets/action_sheet.dart

```diff
@@ -372,37 +391,145 @@ class UserTopicUpdateButton extends ActionSheetMenuItemButton {
+  /// see dartdoc of [ActionSheetMenuItemButton].
+  final int someMessageIdInTopic;
+
+  @override
+  IconData get icon => _actionIsResolve ? ZulipIcons.check : ZulipIcons.check_remove;
+
+  @override
+  String label(ZulipLocalizations zulipLocalizations) {
+    return _actionIsResolve
+      ? zulipLocalizations.actionSheetOptionResolveTopic
+      : zulipLocalizations.actionSheetOptionUnresolveTopic;
+  }
+
+  @override void onPressed() async {
+    final zulipLocalizations = ZulipLocalizations.of(pageContext);
+    final store = PerAccountStoreWidget.of(pageContext);
+    final message = store.messages[someMessageIdInTopic] as StreamMessage?;
+
+    if (message != null && !message.topic.isSameAs(topic)) {
+      // TODO also interrupt on a channel move, with a specific message?
+
+      // The message's topic doesn't match what it was when we opened the
+      // action sheet. So either the resolve/unresolve action was already done,
+      // the topic was renamed, or at least one message in it including
+      // someMessageIdInTopic was moved. In any case, tell the user and don't
+      // make a request.
+      //
+      // This misses the case where some messages in the topic were moved
+      // but not someMessageIdInTopic. Shrug; discussion:
+      //   https://github.com/zulip/zulip-flutter/pull/1301#discussion_r1934565509
+      final title = _actionIsResolve
+        ? zulipLocalizations.resolveTopicInterruptedTitle
+        : zulipLocalizations.unresolveTopicInterruptedTitle;
+      final errorMessage = message.topic.unresolve().isSameAs(topic.unresolve())
+        ? _actionIsResolve
+          ? zulipLocalizations.topicAlreadyResolvedMessage
+          : zulipLocalizations.topicAlreadyUnresolvedMessage
+        // (Might not be a full topic rename; could have been a topic-move of
+        // just some messages. Shrug.)
+        : zulipLocalizations.topicRenamedMessage;
+      showErrorDialog(context: pageContext, title: title, message: errorMessage);
+      return;
+    }
+
+    if (message == null) {
+      // Proceed.
+      //
+      // This happens when [someMessageIdInTopic] came from [Unreads]
+      // and we just haven't fetched the message's details yet
+      // because we haven't opened a message list containing the message.
+      //
+      // Less commonly, this happens if the message was deleted
+      // after the action sheet was opened;
+      // see dartdoc of [ActionSheetMenuItemButton].
+      //
+      // Anyway, we have a message ID, which the server asks for; use it.
+    }
+
+    try {
+      await updateMessage(store.connection,
+        messageId: someMessageIdInTopic,
+        topic: _actionIsResolve ? topic.resolve() : topic.unresolve(),
+        propagateMode: PropagateMode.changeAll,
+        sendNotificationToOldThread: false,
+        sendNotificationToNewThread: true,
+      );
+    } catch (e) {
+      if (!pageContext.mounted) return;
+
+      String? errorMessage;
+
+      switch (e) {
```

Comment:
nit:
```suggestion
      String? errorMessage;
      switch (e) {
```

since this `switch` is the block that sets this variable